src/librustc_codegen_llvm/type_.rs

```diff
@@ -233,7 +234,26 @@ impl Type {
         }
     }
 
-    pub fn ptr_to(&self) -> &Type {
+    pub fn ptr_to(&self, cx: &CodegenCx<'ll, '_>) -> &Type {
+        if self.kind() == TypeKind::Function {
```

Comment:
I really want to avoid us introspecting LLVM types. I think we should either not use non-flat address space pointers *or* we should special-case `fn` pointers - from a quick search, I found only 2 places that create them:
https://github.com/rust-lang/rust/blob/40123a1257f7645bee1c6d3b6eb8c2f6eebd206c/src/librustc_codegen_llvm/type_of.rs#L268
https://github.com/rust-lang/rust/blob/40123a1257f7645bee1c6d3b6eb8c2f6eebd206c/src/librustc_codegen_llvm/meth.rs#L42

Since these are both calling `.ptr_to()` on the result of `FnType::llvm_type`, a helper method on `FnType` can probably be added.